apollo-sqlite-cache/src/main/java/com/apollographql/apollo/cache/normalized/sql/SqlNormalizedCache.kt

```diff
@@ -0,0 +1,97 @@
+package com.apollographql.apollo.cache.normalized.sql
+
+import com.apollographql.apollo.cache.ApolloCacheHeaders.EVICT_AFTER_READ
+import com.apollographql.apollo.cache.CacheHeaders
+import com.apollographql.apollo.cache.normalized.CacheKey
+import com.apollographql.apollo.cache.normalized.NormalizedCache
+import com.apollographql.apollo.cache.normalized.Record
+import com.apollographql.apollo.cache.normalized.RecordFieldJsonAdapter
+import okio.IOException
+
+class SqlNormalizedCache(
+    private val recordFieldAdapter: RecordFieldJsonAdapter,
+    private val cacheQueries: CacheQueries
+) : NormalizedCache() {
+
+  override fun loadRecord(key: String, cacheHeaders: CacheHeaders): Record? {
+    val record = selectRecordForKey(key)
+
+    if (record != null) {
+      if (cacheHeaders.hasHeader(EVICT_AFTER_READ)) {
+        deleteRecord(key)
+      }
+      return record
+    }
+    return nextCache?.loadRecord(key, cacheHeaders)
+  }
+
+  override fun clearAll() {
+    nextCache?.clearAll()
+    cacheQueries.deleteAll()
+  }
+
+  override fun remove(cacheKey: CacheKey, cascade: Boolean): Boolean {
+    val result: Boolean = nextCache?.remove(cacheKey, cascade) ?: false
+    if (result) {
+      return true
+    }
+    return if (cascade) {
+      selectRecordForKey(cacheKey.key)
+          ?.referencedFields()
+          ?.all { remove(CacheKey(it.key()), cascade = true) }
+          ?: false
+    } else {
+      deleteRecord(cacheKey.key)
+    }
+  }
+
+  override fun performMerge(apolloRecord: Record, cacheHeaders: CacheHeaders): Set<String> {
+    val oldRecord = selectRecordForKey(apolloRecord.key())
+    return if (oldRecord == null) {
+      cacheQueries.insert(apolloRecord.key(), recordFieldAdapter.toJson(apolloRecord.fields()))
+      emptySet()
+    } else {
+      oldRecord.mergeWith(apolloRecord).also {
+        if (it.isNotEmpty()) {
+          cacheQueries.update(oldRecord.key(), recordFieldAdapter.toJson(oldRecord.fields()))
+        }
+      }
+    }
+  }
+
+  fun selectRecordForKey(key: String): Record? {
+    return try {
+      cacheQueries.recordForKey(key)
+          .executeAsList()
+          .firstOrNull()
```

Comment:
```suggestion
          .executeAsOneOrNull()
```

Reply:
I remember there was a specific reason why I chose asList. I will double check.

Reply:
Ok, I see now. `executeAsOneOrNull` is more like `singleOrNull` function in Kotlin. If there are more, it crashes. The old logic was going to over the Cursor and picks the first match too. Since `key` in the database is not `UNIQUE`, we may have multiple of them.

Not sure if we should though 😄 Don't really want to change the behavior now. It can be something we may look in the future.

Reply:
Makes senses 👍